File: src/main/java/net/onelitefeather/pandorascluster/service/LandEntityListener.java
package net.onelitefeather.pandorascluster.service;

import com.destroystokyo.paper.event.block.TNTPrimeEvent;
import com.destroystokyo.paper.event.entity.EntityPathfindEvent;
import net.onelitefeather.pandorascluster.enums.Permission;
import net.onelitefeather.pandorascluster.land.Land;
import net.onelitefeather.pandorascluster.land.flag.LandFlag;
import net.onelitefeather.pandorascluster.land.flag.LandFlagEntity;
import net.onelitefeather.pandorascluster.service.services.LandFlagService;
import org.bukkit.GameMode;
import org.bukkit.Location;
import org.bukkit.block.Block;
import org.bukkit.block.data.Powerable;
import org.bukkit.block.data.type.Farmland;
import org.bukkit.entity.*;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.block.EntityBlockFormEvent;
import org.bukkit.event.entity.*;
import org.bukkit.event.hanging.HangingBreakEvent;
import org.bukkit.event.vehicle.*;
import org.bukkit.inventory.ItemStack;
import org.bukkit.permissions.Permissible;
import org.bukkit.projectiles.ProjectileSource;
import org.jetbrains.annotations.NotNull;
import org.spigotmc.event.entity.EntityMountEvent;

import java.util.Iterator;
import java.util.List;

record LandEntityListener(LandService landService, LandFlagService landFlagService) implements Listener {

    @EventHandler
    public void theConfuser(EntityDamageByEntityEvent event) {

        var target = event.getEntity();
        var attacker = event.getDamager();

        var land = this.landService.getLand(target.getChunk());

        if (land == null) land = this.landService.getLand(attacker.getChunk());

        if (land == null) return;
        event.setCancelled(true);
    }

    @EventHandler
    public void handleEntityExplode(TNTPrimeEvent event) {

        var block = event.getBlock();
        var land = this.landService.getLand(block.getChunk());
        var primerEntity = event.getPrimerEntity();

        if (land != null) {

            LandFlagEntity landFlag = this.landFlagService.getFlag(LandFlag.EXPLOSIONS, land);
            var cancel = landFlag != null && !landFlag.<Boolean>getValue();

            if (primerEntity != null) {
                if (land.hasAccess(primerEntity.getUniqueId())) return;
                if (Permission.EXPLOSION.hasPermission(primerEntity)) return;
                cancel = true;
            }

            event.setCancelled(cancel);
        }
    }

    @EventHandler
    public void handleEntityExplode(EntityExplodeEvent event) {

        var entity = event.getEntity();
        var land = this.landService.getLand(entity.getChunk());

        if (land != null) {

            Iterator<Block> iterator = event.blockList().iterator();
            while (iterator.hasNext()) {

                var next = iterator.next();
                var nextLand = this.landService.getLand(next.getChunk());

                if (nextLand != null && !this.landService.hasSameOwner(land, nextLand)) {
                    iterator.remove();
                }
            }
        }
    }

    @EventHandler
    public void handleEntityMount(EntityMountEvent event) {

        var mount = event.getMount();
        var entity = event.getEntity();

        if (!(mount instanceof Vehicle)) return;

        if (mount instanceof Tameable tameable) {
            var tamer = tameable.getOwner();
            if (tameable.isTamed() && tamer != null && tamer.getUniqueId().equals(entity.getUniqueId())) return;
        }

        var land = this.landService.getLand(mount.getChunk());
        if (land == null) return;

        if (land.hasAccess(entity.getUniqueId())) return;
        if (Permission.ENTITY_MOUNT.hasPermission(entity)) return;
        event.setCancelled(true);
    }

    @EventHandler
    public void handleEntityChangeBlock(EntityChangeBlockEvent event) {
        var land = this.landService.getLand(event.getBlock().getChunk());

        if (land == null) return;

        var landFlag = this.landFlagService.getFlag(LandFlag.BLOCK_FORM, land);
        if (landFlag != null && landFlag.<Boolean>getValue()) return;
        event.setCancelled(true);
    }

    @EventHandler
    public void handleEntityBlockForm(EntityBlockFormEvent event) {

        var land = this.landService.getLand(event.getBlock().getChunk());
        if (land == null) return;

        var landFlag = this.landFlagService.getFlag(LandFlag.BLOCK_FORM, land);
        if (landFlag != null) {
            if (land.hasAccess(event.getEntity().getUniqueId())) return;

            boolean flag = landFlag.getValue();
            if (flag) return;

            event.setCancelled(true);
        }
    }

    @EventHandler
    public void handleHangingBreak(HangingBreakEvent event) {

        var entity = event.getEntity();

        var land = this.landService.getLand(entity.getChunk());
        if (land == null) return;

        var landFlag = this.landFlagService.getFlag(LandFlag.HANGING_BREAK, land);
        if (landFlag != null && landFlag.<Boolean>getValue()) return;
        event.setCancelled(true);
    }

    @EventHandler
    public void handleEntityInteract(EntityInteractEvent event) {

        var entity = event.getEntity();
        var block = event.getBlock();

        var land = this.landService.getLand(block.getChunk());
        if (land == null) return;

        var data = block.getBlockData();

        var farmLandDestroyFlag = this.landFlagService.getFlag(LandFlag.FARMLAND_DESTROY, land);
        var redstoneFlag = this.landFlagService.getFlag(LandFlag.REDSTONE, land);

        if (data instanceof Farmland && farmLandDestroyFlag != null && !farmLandDestroyFlag.<Boolean>getValue() ||
                data instanceof Powerable && redstoneFlag != null && !redstoneFlag.<Boolean>getValue()) {
            event.setCancelled(true);
        }

    }

    @EventHandler
    public void handleEntityPathfinding(EntityPathfindEvent event) {

        var entity = event.getEntity();

        var location = event.getLoc();
        var land = this.landService.getLand(location.getChunk());
        var entityLand = this.landService.getLand(entity.getChunk());

        var cancel = false;

        if (entityLand != null) {
            if (land != null) {
                if (!this.landService.hasSameOwner(entityLand, land)) {
                    cancel = true;
                }
            } else {
                cancel = true;
            }
        } else {
            if (land != null) {
                cancel = true;
            }
        }

        event.setCancelled(cancel);
    }

    @EventHandler
    public void handleEntityTargetLivingEntity(EntityTargetLivingEntityEvent event) {

        var entity = event.getEntity();
        var target = event.getTarget();

        var land = this.landService.getLand(entity.getChunk());
        if (target != null && land == null) {
            land = this.landService.getLand(target.getChunk());
        }

        if (land == null) return;

        var landFlag = this.landFlagService.getFlag(LandFlag.PVE, land);
        if (landFlag != null && landFlag.<Boolean>getValue()) return;
        event.setCancelled(true);
    }

    @EventHandler
    public void handleVehicleEnter(VehicleEnterEvent event) {

        var vehicle = event.getVehicle();
        var entered = event.getEntered();

        var land = this.landService.getLand(vehicle.getChunk());

        if (land != null) {

            LandFlagEntity landFlag = this.landFlagService.getFlag(LandFlag.VEHICLE_USE, land);
            if (landFlag != null && landFlag.<Boolean>getValue()) return;

            if (land.hasAccess(entered.getUniqueId())) return;

            if (Permission.VEHICLE_ENTER.hasPermission(entered)) return;
            event.setCancelled(true);
        }
    }

    @EventHandler
    public void handleVehicleDestroy(VehicleDestroyEvent event) {

        var vehicle = event.getVehicle();
        var attacker = event.getAttacker();

        var land = this.landService.getLand(vehicle.getChunk());

        if (land != null) {

            var landFlag = this.landFlagService.getFlag(LandFlag.VEHICLE_DAMAGE, land);
            if (landFlag != null && landFlag.<Boolean>getValue()) return;

            if (attacker != null) {
                if (land.hasAccess(attacker.getUniqueId())) return;
                if (Permission.VEHICLE_DESTROY.hasPermission(attacker)) return;
                event.setCancelled(true);
                return;
            }

            event.setCancelled(true);
        }
    }

    @EventHandler
    public void handleVehicleDamage(VehicleDamageEvent event) {

        var vehicle = event.getVehicle();
        var attacker = event.getAttacker();

        var land = this.landService.getLand(vehicle.getChunk());
        if (land != null) {

            if (attacker != null) {
                if (land.hasAccess(attacker.getUniqueId())) return;
                if (Permission.VEHICLE_DAMAGE.hasPermission(attacker)) return;
            }

            var landFlag = this.landFlagService.getFlag(LandFlag.VEHICLE_DAMAGE, land);
            if (landFlag != null && landFlag.<Boolean>getValue()) return;
            event.setCancelled(true);
        }
    }

    @EventHandler
    public void handleVehicleCreation(VehicleCreateEvent event) {

        var vehicle = event.getVehicle();
        var land = this.landService.getLand(vehicle.getChunk());
        if (land != null) {
            var landFlag = this.landFlagService.getFlag(LandFlag.VEHICLE_CREATE, land);
            if (landFlag != null && landFlag.<Boolean>getValue()) return;
            event.setCancelled(true);
        }
    }

    @EventHandler
    public void handleVehicleMovement(VehicleMoveEvent event) {

        var vehicle = event.getVehicle();

        var to = event.getTo();
        var from = event.getFrom();

        var toLand = this.landService.getLand(to.getChunk());
        var fromLand = this.landService.getLand(from.getChunk());

        if (toLand != null) {
            for (var entity : vehicle.getPassengers()) {
                if (toLand.isBanned(entity.getUniqueId())) {
                    vehicle.removePassenger(entity);
                }
            }
        }

        if (fromLand != null) {
            removeVehicle(vehicle, toLand != null && !this.landService.hasSameOwner(toLand, fromLand) ? to : from);
        }
    }

    @EventHandler(priority = EventPriority.HIGHEST)
    public void handlePotionSplash(ProjectileLaunchEvent event) {

        var projectile = event.getEntity();

        if (projectile instanceof ThrownPotion thrownPotion) {

            var source = thrownPotion.getShooter();

            if (source instanceof Entity entity) {
                var land = this.landService.getLand(entity.getChunk());

                if (land != null) {

                    var landFlag = this.landFlagService.getFlag(LandFlag.POTION_SPLASH, land);
                    if (landFlag != null && landFlag.<Boolean>getValue()) return;

                    if (land.hasAccess(entity.getUniqueId())) return;
                    if (Permission.POTION_SPLASH.hasPermission(entity)) return;

                    event.setCancelled(true);
                }
            }
        }
    }

    @EventHandler
    public void handleProjectileHit(ProjectileHitEvent event) {

        var entity = event.getEntity();
        ProjectileSource shooter = null;

        if (entity.getShooter() instanceof Entity) {
            shooter = entity.getShooter();
        }

        var hitEntity = event.getHitEntity();
        var hitBlock = event.getHitBlock();

        Land land = null;

        if (hitEntity != null) {
            land = this.landService.getLand(hitEntity.getChunk());
        }

        if (hitBlock != null) {
            land = this.landService.getLand(hitBlock.getChunk());
        }

        if (land != null) {

            var cancel = false;

            if (shooter instanceof Permissible permissible) {
                cancel = !Permission.PROJECTILE_HIT_ENTITY.hasPermission(permissible);
            }


            if (hitBlock != null) {
                var blockData = hitBlock.getBlockData();
                if (blockData instanceof Powerable) {
                    var landFlag = this.landFlagService.getFlag(LandFlag.REDSTONE, land);
                    cancel = landFlag != null && !landFlag.<Boolean>getValue();
                }
            }

            event.setCancelled(cancel);
        }
    }

    private void removeVehicle(@NotNull Vehicle vehicle, @NotNull Location location) {
        List<Entity> passengers = vehicle.getPassengers();

        boolean cancelDrop = false;
        if (!passengers.isEmpty()) {
            for (int i = 0; i < passengers.size() && !cancelDrop; i++) {
                var passenger = passengers.get(i);
                if (passenger instanceof Player player && player.getGameMode() == GameMode.CREATIVE) {
                    cancelDrop = true;
                }
            }
        }

        if (!cancelDrop) {

            ItemStack dropItem = null;
            if (vehicle instanceof Minecart minecart) {
                dropItem = new ItemStack(minecart.getMinecartMaterial());
            } else if (vehicle instanceof Boat boat) {
                dropItem = new ItemStack(boat.getBoatType().getMaterial());
            }

            if (dropItem != null) {
                location.getWorld().dropItemNaturally(location, dropItem);
            }
        }

        vehicle.remove();
    }

}

File: src/main/kotlin/net/onelitefeather/pandorascluster/listener/LandEntityListener.kt
package net.onelitefeather.pandorascluster.listener

import com.destroystokyo.paper.event.block.TNTPrimeEvent
import net.onelitefeather.pandorascluster.api.PandorasClusterApi
import net.onelitefeather.pandorascluster.enums.Permission
import net.onelitefeather.pandorascluster.extensions.hasPermission
import net.onelitefeather.pandorascluster.land.flag.LandFlag
import org.bukkit.block.data.type.Farmland
import org.bukkit.block.data.type.TurtleEgg
import org.bukkit.entity.*
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.entity.*
import org.bukkit.event.vehicle.*
import org.bukkit.permissions.Permissible
import org.spigotmc.event.entity.EntityMountEvent

class LandEntityListener(private val pandorasClusterApi: PandorasClusterApi) :
    Listener {

    @EventHandler
    fun handleProjectileBlockHit(event: ProjectileHitEvent) {
        val entity = event.entity
        val shooter = if (entity.shooter is Entity) {
            entity.shooter as Entity
        } else null ?: return
        val hitBlock = event.hitBlock
        if (hitBlock != null) {
            val land = pandorasClusterApi.getLand(hitBlock.chunk) ?: return
            event.isCancelled = !land.hasAccess(shooter.uniqueId)
        }
    }

    @EventHandler
    fun theConfuser(event: EntityDamageByEntityEvent) {

        val target = event.entity
        var attacker = event.damager

        if (attacker is Projectile) {
            attacker = attacker.shooter as Entity
        }

        val land = pandorasClusterApi.getLand(target.chunk) ?: pandorasClusterApi.getLand(attacker.chunk) ?: return

        event.isCancelled = if (target is Player && attacker is Player) {
            val flag = pandorasClusterApi.getLandFlag(LandFlag.PVP, land) ?: return
            val value = flag.getValue<Boolean>() == true
            if (attacker.hasPermission(Permission.PVP)) return
            if (land.hasAccess(attacker.uniqueId) && land.hasAccess(target.uniqueId)) return
            !value
        } else {
            val flag = pandorasClusterApi.getLandFlag(LandFlag.PVE, land) ?: return
            val value = flag.getValue<Boolean>() == true
            if (target.hasPermission(Permission.PVE) || attacker.hasPermission(Permission.PVE)) return
            if (land.hasAccess(target.uniqueId) || land.hasAccess(attacker.uniqueId)) return
            !value
        }
    }

    @EventHandler
    fun handleEntityExplode(event: TNTPrimeEvent) {
        val block = event.block
        val land = pandorasClusterApi.getLand(block.chunk)
        val primerEntity = event.primerEntity
        if (land != null) {
            val landFlag = land.getLandFlag(LandFlag.EXPLOSIONS)
            event.isCancelled = if (landFlag.getValue<Boolean>() == false) {
                true
            } else if (primerEntity != null) {
                if (land.hasAccess(primerEntity.uniqueId)) return
                if (primerEntity.hasPermission(Permission.EXPLOSION)) return
                true
            } else false
        }
    }

    @EventHandler
    fun handleEntityExplode(event: EntityExplodeEvent) {
        event.blockList().groupBy { it.chunk }.filter {
            pandorasClusterApi.getLand(it.key)?.getLandFlag(LandFlag.EXPLOSIONS)?.getValue<Boolean>() == false
        }.forEach { event.blockList().removeAll(it.value) }
    }

    @EventHandler
    fun handleEntityTargetLivingEntity(event: EntityTargetLivingEntityEvent) {
        val target = event.target ?: return
        val land = pandorasClusterApi.getLand(target.chunk) ?: return
        val landFlag = pandorasClusterApi.getLandFlag(LandFlag.PVE, land) ?: return
        event.isCancelled = landFlag.getValue<Boolean>() == true
    }

    @EventHandler
    fun handleEntityInteract(event: EntityInteractEvent) {

        val block = event.block
        val land = pandorasClusterApi.getLand(block.chunk) ?: return
        val blockData = block.blockData

        if (blockData is TurtleEgg) {
            val turtleEggDestroyFlag = pandorasClusterApi.getLandFlag(LandFlag.TURTLE_EGG_DESTROY, land)
            if (turtleEggDestroyFlag != null && turtleEggDestroyFlag.getValue<Boolean>() == false) {
                event.isCancelled = true
                event.entity.velocity =
                    event.entity.velocity.subtract(event.entity.location.direction).multiply(0.5).normalize()
                return
            }
        }

        val farmLandDestroyFlag = pandorasClusterApi.getLandFlag(LandFlag.FARMLAND_DESTROY, land) ?: return
        event.isCancelled = blockData is Farmland && farmLandDestroyFlag.getValue<Boolean>() == false
    }

    @EventHandler
    fun handleEntityMount(event: EntityMountEvent) {

        val mount = event.mount
        val entity = event.entity

        if (mount !is Vehicle) return
        if (mount is Tameable) {
            val tamer = mount.owner
            if (mount.isTamed && tamer != null && tamer.uniqueId == entity.uniqueId) return
        }

        val land = this.pandorasClusterApi.getLand(mount.getChunk()) ?: return
        if (land.hasAccess(entity.uniqueId)) return
        if (Permission.ENTITY_MOUNT.hasPermission(entity)) return
        event.isCancelled = true
    }

    @EventHandler
    fun handleEntityChangeBlock(event: EntityChangeBlockEvent) {
        val land = pandorasClusterApi.getLand(event.block.chunk) ?: return
        val landFlag = pandorasClusterApi.getLandFlag(LandFlag.ENTITY_CHANGE_BLOCK, land)
        event.isCancelled = landFlag?.getValue<Boolean>() == false
    }

    @EventHandler
    fun handleVehicleDestroy(event: VehicleDestroyEvent) {
        val vehicle = event.vehicle
        val attacker = event.attacker
        val land = pandorasClusterApi.getLand(vehicle.chunk)
        if (land != null) {
            val landFlag = pandorasClusterApi.getLandFlag(LandFlag.VEHICLE_DAMAGE, land) ?: return
            if (landFlag.getValue<Boolean>() == true) return
            if (attacker != null) {
                if (land.hasAccess(attacker.uniqueId)) return
                if (Permission.VEHICLE_DESTROY.hasPermission(attacker)) return
                event.isCancelled = true
                return
            }
            event.isCancelled = true
        }
    }

    @EventHandler
    fun handleVehicleDamage(event: VehicleDamageEvent) {
        val vehicle = event.vehicle
        val attacker = event.attacker
        val land = pandorasClusterApi.getLand(vehicle.chunk)
        if (land != null) {
            if (attacker != null) {
                if (land.hasAccess(attacker.uniqueId)) return
                if (Permission.VEHICLE_DAMAGE.hasPermission(attacker)) return
            }
            val landFlag = pandorasClusterApi.getLandFlag(LandFlag.VEHICLE_DAMAGE, land) ?: return
            if (landFlag.getValue<Boolean>() == true) return
            event.isCancelled = true
        }
    }

    @EventHandler
    fun handleVehicleCreation(event: VehicleCreateEvent) {
        val vehicle = event.vehicle
        val land = pandorasClusterApi.getLand(vehicle.chunk) ?: return
        val landFlag = pandorasClusterApi.getLandFlag(LandFlag.VEHICLE_CREATE, land) ?: return
        if (landFlag.getValue<Boolean>() == true) return
        event.isCancelled = true
    }

    @EventHandler
    fun handleVehicleMovement(event: VehicleMoveEvent) {
        val vehicle = event.vehicle
        val to = event.to
        val toLand = pandorasClusterApi.getLand(to.chunk)
        if (toLand != null) {
            vehicle.passengers.filter { toLand.isBanned(it.uniqueId) }.forEach { vehicle.removePassenger(it) }
        } else if (vehicle is Minecart || vehicle is Boat) {
            vehicle.remove()
        }

    }

    @Suppress("kotlin:S3776")
    @EventHandler(priority = EventPriority.HIGHEST)
    fun handlePotionSplash(event: ProjectileLaunchEvent) {

        val projectile = event.entity

        if (projectile is ThrownPotion) {

            val source = projectile.getShooter()
            if (source is Entity) {
                val land = pandorasClusterApi.getLand(source.chunk)

                if (land != null) {

                    val landFlag = pandorasClusterApi.getLandFlag(LandFlag.POTION_SPLASH, land) ?: return
                    if (landFlag.getValue<Boolean>() == true) return

                    if (land.hasAccess(projectile.uniqueId)) return
                    if (Permission.POTION_SPLASH.hasPermission(projectile)) return
                    event.isCancelled = true
                }
            }
        }
    }

    @EventHandler
    fun handleVehicleEnter(event: VehicleEnterEvent) {
        val vehicle = event.vehicle
        val entered = event.entered
        val land = pandorasClusterApi.getLand(vehicle.chunk)
        if (land != null) {
            val landFlag = pandorasClusterApi.getLandFlag(LandFlag.VEHICLE_USE, land)
            if (landFlag != null && landFlag.getValue<Boolean>() == true) return
            if (land.hasAccess(entered.uniqueId)) return
            if (Permission.VEHICLE_ENTER.hasPermission(entered)) return
            event.isCancelled = true
        }
    }

    @EventHandler
    fun handleEntityTame(event: EntityTameEvent) {
        val land = pandorasClusterApi.getLand(event.entity.chunk)
        val owner = event.owner
        if (owner !is Permissible) return
        if (Permission.TAME_ENTITY.hasPermission(owner)) return
        if (land == null || !land.isBanned(event.owner.uniqueId)) return
        event.isCancelled = true
    }

    @EventHandler
    fun handleEntityEnterBlock(event: EntityEnterBlockEvent) {
        val land = pandorasClusterApi.getLand(event.block.chunk) ?: return
        val landFlag = pandorasClusterApi.getLandFlag(LandFlag.BEE_INTERACT, land)
        event.isCancelled = landFlag?.getValue<Boolean>() == false
    }
}
